contracts/stake-cw20-external-rewards/src/contract.rs

```diff
@@ -0,0 +1,1324 @@
+    }
+}
+
+pub fn try_fund(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    let mut config = CONFIG.load(deps.storage)?;
+    if config.funded {
+        return Err(ContractError::AlreadyFunded {});
+    }
+    if config.denom != Denom::Cw20(info.sender) {
+        return Err(ContractError::IncorrectDenom {});
+    }
+    if amount != config.total_amount {
+        return Err(ContractError::IncorrectFundingAmount {
+            received: amount,
+            expected: config.total_amount,
+        });
+    }
+    config.funded = true;
+    CONFIG.save(deps.storage, &config)?;
+    Ok(Response::new()
+        .add_attribute("action", "fund")
+        .add_attribute("amount", amount))
+}
+
+pub fn try_claim(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    up_to_block: u64,
+) -> Result<Response, ContractError> {
+    if up_to_block > env.block.height {
+        return Err(ContractError::InvalidFutureBlock {});
+    }
+    let config = CONFIG.load(deps.storage)?;
+    if !config.funded {
+        return Err(ContractError::NotFunded {});
+    };
+    if env.block.height < config.start_block {
+        return Err(RewardsNotStarted {
+            current_block: env.block.height,
+            start_block: config.start_block,
+        });
+    }
+    let last_claimed = LAST_CLAIMED
+        .may_load(deps.storage, info.sender.clone())?
+        .unwrap_or(LastClaim {
+            block_height: config.start_block,
+            time: Timestamp::default(),
+        });
+    let (amount_owed, new_claim_height) = get_amount_owed(
+        deps.as_ref(),
+        &info.sender,
+        &config,
+        last_claimed.block_height,
+        up_to_block,
+    )?;
+
+    if amount_owed == Uint128::zero() {
+        return Err(ContractError::ZeroClaimable {});
+    }
+
+    let new_claim = LastClaim {
+        block_height: new_claim_height,
+        time: env.block.time,
+    };
+    LAST_CLAIMED.save(deps.storage, info.sender.clone(), &new_claim)?;
+
+    let payment_msg = get_payment_msg(&env, &config, &info, amount_owed)?;
+
+    Ok(Response::new()
+        .add_message(payment_msg)
+        .add_attribute("action", "claim")
+        .add_attribute("amount", amount_owed))
+}
+
+fn get_amount_owed(
+    deps: Deps,
+    address: &Addr,
+    config: &Config,
+    last_claimed: u64,
+    up_to_block: u64,
+) -> StdResult<(Uint128, u64)> {
+    let delta = config.payment_block_delta;
+    let mut current_block = last_claimed;
+    let mut amount = Uint128::zero();
+    while current_block <= min(up_to_block, config.end_block) {
+        let stake_info = get_stake_info_at_height(deps.querier, config, address, current_block)?;
+        if stake_info.balance > Uint128::zero() {
+            amount += ((Uint128::from(delta) * config.payment_per_block) * stake_info.balance)
+                .checked_div(stake_info.total)?
```

Comment:
This div will round down so unless things divide pretty nicely this will end up distributing slightly (on order of `10 ** -6 juno`) less rewards than it has been given I believe.

Probably fine but I wonder if you have thoughts about that. I've got the same issue with the DAO Up! contracts.

Reply:
yeah, i think rounding down is acceptable. Not sure there is any other option. I guess you could add a refund method for the dust at the end but idk if it really matters for most sane tokens. I could see this being an issue for tokens with really low supply and no decimals for example :man_shrugging:

Reply:
This is the least of our worries IMO. Dust is just a part of crypto...